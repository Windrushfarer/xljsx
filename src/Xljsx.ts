import {
  CellElement,
  RowElement,
  WorksheetElement,
  WorkbookElement,
  HeaderElement,
  RowGroupElement,
  ColumnElement,
  IWorksheetAttributes,
  IWorkbookAttributes,
  IRowGroupAttributes,
  IColumnAttributes,
  IHeaderAttributes,
  IRowAttributes,
  ICellAttributes,
} from './intrinsics'

import { Node, FC } from './interfaces'

type AllElements =
  | CellElement
  | RowElement
  | WorksheetElement
  | WorkbookElement
  | HeaderElement
  | RowGroupElement
  | ColumnElement
  | null

type IntrinsicsAttributes =
  | ICellAttributes
  | IRowAttributes
  | IHeaderAttributes
  | IColumnAttributes
  | IRowGroupAttributes
  | IWorksheetAttributes
  | IWorkbookAttributes

function isElement(element: any): element is AllElements {
  return (
    CellElement.isCellElement(element) ||
    ColumnElement.isColumnElement(element) ||
    RowElement.isRowElement(element) ||
    RowGroupElement.isRowGroupElement(element) ||
    HeaderElement.isHeaderElement(element) ||
    WorksheetElement.isWorksheetElement(element) ||
    WorkbookElement.isWorkbookElement(element)
  )
}

function create<P = {}>(type: FC<P>, attributes: P | null, ...children: Node[]): AllElements
function create(type: 'cell', attributes: ICellAttributes, ...children: Node[]): CellElement
function create(type: 'header', attributes: IHeaderAttributes | null, ...children: Node[]): HeaderElement
function create(type: 'column', attributes: IColumnAttributes, ...children: Node[]): ColumnElement
function create(type: 'row', attributes: IRowAttributes | null, ...children: Node[]): RowElement
function create(type: 'rowGroup', attributes: IRowGroupAttributes | null, ...children: Node[]): RowGroupElement
function create(type: 'worksheet', attributes: IWorksheetAttributes, ...children: Node[]): WorksheetElement
function create(type: 'workbook', attributes: IWorkbookAttributes | null, ...children: Node[]): WorkbookElement
function create<P = {}>(
  type: 'cell' | 'row' | 'worksheet' | 'workbook' | 'header' | 'rowGroup' | 'column' | FC<P>,
  attributes: P | IntrinsicsAttributes | null,
  ...rawChildren: Node[]
): AllElements {
  const children = rawChildren

  if (typeof type === 'function') {
    const attrs = attributes ? (attributes as P) : null
    const props = attrs
      ? Object.assign(attrs, {
          children,
        })
      : ({ children } as P & { children: Node[] })

    const result = type(props)

    if (isElement(result)) {
      return result
    }

    return null
  }

  switch (type) {
    case 'cell': {
      return new CellElement(attributes as ICellAttributes, children)
    }
    case 'row': {
      return new RowElement((attributes || {}) as IRowAttributes, children)
    }
    case 'header': {
      return new HeaderElement((attributes || {}) as IHeaderAttributes, children)
    }
    case 'column': {
      return new ColumnElement(attributes as IColumnAttributes, children)
    }
    case 'rowGroup': {
      return new RowGroupElement((attributes || {}) as IRowGroupAttributes, children)
    }
    case 'worksheet': {
      return new WorksheetElement(attributes as IWorksheetAttributes, children)
    }
    case 'workbook': {
      return new WorkbookElement((attributes || {}) as IWorkbookAttributes, children)
    }
    default: {
      return null
    }
  }
}

export default {
  create,
}
