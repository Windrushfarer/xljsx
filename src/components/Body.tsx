import Xljsx from '../Xljsx'

import { FC } from '../interfaces'

interface IProps {
  cellId?: 'a'
}

const Body: FC<IProps> = (props) => {
  const { cellId, children } = props

  return (
    <row>
      <cell id={cellId || 'a'}>A</cell>
      {children}
    </row>
  )
}

export default Body
