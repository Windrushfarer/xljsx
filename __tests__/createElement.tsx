import Xljsx from '../src/Xljsx'
import Body from '../src/components/Body'

describe('function component', () => {
  it('should works', () => {
    const cell = Xljsx.create('cell', { id: 'B' }, 'something', 'wrong', false, 'troe')
    const bodyOne = Xljsx.create(Body, null, cell)

    expect(bodyOne).toMatchSnapshot()
  })
})

describe('cells', () => {
  it('should create cell', () => {
    const cell = Xljsx.create('cell', { id: 'A' }, 'something', 'wrong', false, 'troe')

    expect(cell).toMatchSnapshot()

    expect(
      <cell id={'A'}>
        something{'wrong'}
        {false}troe
      </cell>
    ).toEqual(cell)
  })
})

describe('rows', () => {
  it('should create row with cell', () => {
    const cell = Xljsx.create('cell', { id: 'A' }, 'something')
    const row = Xljsx.create('row', null, cell)

    expect(row).toMatchSnapshot()

    expect(
      <row>
        <cell id={'A'}>something</cell>
      </row>
    ).toEqual(row)
  })
})

describe('rowGroups', () => {
  it('should create rowGroup with rows', () => {
    const cellA = Xljsx.create('cell', { id: 'A' }, 'A')
    const cellB = Xljsx.create('cell', { id: 'B' }, 'B')
    const cellC = Xljsx.create('cell', { id: 'C' }, 'C')

    const rowOne = Xljsx.create('row', null, cellA, cellB, cellC)
    const rowTwo = Xljsx.create('row', null, cellA, cellB, cellC)

    const rowGroup = Xljsx.create('rowGroup', null, rowOne, rowTwo)

    expect(rowGroup).toMatchSnapshot()

    expect(
      <rowGroup>
        <row>
          <cell id="A">A</cell>
          <cell id="B">B</cell>
          <cell id="C">C</cell>
        </row>
        <row>
          <cell id="A">A</cell>
          <cell id="B">B</cell>
          <cell id="C">C</cell>
        </row>
      </rowGroup>
    ).toEqual(rowGroup)
  })
})

describe('worksheets', () => {
  it('should create worksheet with row and cells', () => {
    const hColumnA = Xljsx.create('column', { id: 'A' }, 'Column A')
    const hColumnB = Xljsx.create('column', { id: 'B' }, 'Column B')
    const hColumnC = Xljsx.create('column', { id: 'C' }, 'Column C')

    const header = Xljsx.create('header', null, hColumnA, hColumnB, hColumnC)

    const cellA = Xljsx.create('cell', { id: 'A' }, 'CellA')
    const cellB = Xljsx.create('cell', { id: 'B' }, 'CellB')
    const cellC = Xljsx.create('cell', { id: 'C' }, 'CellC')

    const row = Xljsx.create('row', null, cellA, cellB, cellC)

    const worksheet = Xljsx.create('worksheet', { name: 'Example' }, header, row)

    expect(worksheet).toMatchSnapshot()

    expect(
      <worksheet name="Example">
        <header>
          <column id={'A'}>Column A</column>
          <column id={'B'}>Column B</column>
          <column id={'C'}>Column C</column>
        </header>
        <row>
          <cell id={'A'}>CellA</cell>
          <cell id={'B'}>CellB</cell>
          <cell id={'C'}>CellC</cell>
        </row>
      </worksheet>
    ).toEqual(worksheet)
  })
})

describe('workbooks', () => {
  it('should create workbook with worksheets', () => {
    const hColumnA = Xljsx.create('column', { id: 'A' }, 'Column A')
    const hColumnB = Xljsx.create('column', { id: 'B' }, 'Column B')
    const hColumnC = Xljsx.create('column', { id: 'C' }, 'Column C')

    const cellA = Xljsx.create('cell', { id: 'A' }, 'CellA')
    const cellB = Xljsx.create('cell', { id: 'B' }, 'CellB')
    const cellC = Xljsx.create('cell', { id: 'C' }, 'CellC')

    const header = Xljsx.create('header', null, hColumnA, hColumnB, hColumnC)

    const row = Xljsx.create('row', null, cellA, cellB, cellC)

    const worksheet = Xljsx.create('worksheet', { name: 'Example' }, header, row)
    const worksheet2 = Xljsx.create('worksheet', { name: 'Example2' }, header, row)

    const workbook = Xljsx.create(
      'workbook',
      { creator: 'Alexander Simonov' },
      worksheet,
      worksheet2
    )

    expect(workbook).toMatchSnapshot()

    expect(
      <workbook creator="Alexander Simonov">
        <worksheet name="Example">
          <header>
            <column id={'A'}>Column A</column>
            <column id={'B'}>Column B</column>
            <column id={'C'}>Column C</column>
          </header>
          <row>
            <cell id={'A'}>CellA</cell>
            <cell id={'B'}>CellB</cell>
            <cell id={'C'}>CellC</cell>
          </row>
        </worksheet>
        <worksheet name="Example2">
          <header>
            <column id={'A'}>Column A</column>
            <column id={'B'}>Column B</column>
            <column id={'C'}>Column C</column>
          </header>
          <row>
            <cell id={'A'}>CellA</cell>
            <cell id={'B'}>CellB</cell>
            <cell id={'C'}>CellC</cell>
          </row>
        </worksheet>
      </workbook>
    ).toEqual(workbook)
  })
})
