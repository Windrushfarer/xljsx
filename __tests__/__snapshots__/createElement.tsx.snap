// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cells should create cell 1`] = `
CellElement {
  "options": Object {
    "id": "A",
  },
  "type": "cell",
  "value": "somethingwrongtroe",
}
`;

exports[`function component should works 1`] = `
RowElement {
  "cells": Array [
    CellElement {
      "options": Object {
        "id": "a",
      },
      "type": "cell",
      "value": "A",
    },
    CellElement {
      "options": Object {
        "id": "B",
      },
      "type": "cell",
      "value": "somethingwrongtroe",
    },
  ],
  "options": Object {},
  "type": "row",
}
`;

exports[`rowGroups should create rowGroup with rows 1`] = `
RowGroupElement {
  "options": Object {},
  "rowsAndGroups": Array [
    RowElement {
      "cells": Array [
        CellElement {
          "options": Object {
            "id": "A",
          },
          "type": "cell",
          "value": "A",
        },
        CellElement {
          "options": Object {
            "id": "B",
          },
          "type": "cell",
          "value": "B",
        },
        CellElement {
          "options": Object {
            "id": "C",
          },
          "type": "cell",
          "value": "C",
        },
      ],
      "options": Object {},
      "type": "row",
    },
    RowElement {
      "cells": Array [
        CellElement {
          "options": Object {
            "id": "A",
          },
          "type": "cell",
          "value": "A",
        },
        CellElement {
          "options": Object {
            "id": "B",
          },
          "type": "cell",
          "value": "B",
        },
        CellElement {
          "options": Object {
            "id": "C",
          },
          "type": "cell",
          "value": "C",
        },
      ],
      "options": Object {},
      "type": "row",
    },
  ],
  "type": "rowGroup",
}
`;

exports[`rows should create row with cell 1`] = `
RowElement {
  "cells": Array [
    CellElement {
      "options": Object {
        "id": "A",
      },
      "type": "cell",
      "value": "something",
    },
  ],
  "options": Object {},
  "type": "row",
}
`;

exports[`workbooks should create workbook with worksheets 1`] = `
Workbook {
  "options": Object {
    "creator": "Alexander Simonov",
  },
  "type": "workbook",
  "worksheets": Array [
    WorksheetElement {
      "header": HeaderElement {
        "columns": Array [
          ColumnElement {
            "name": "Column A",
            "options": Object {
              "id": "A",
            },
            "type": "column",
          },
          ColumnElement {
            "name": "Column B",
            "options": Object {
              "id": "B",
            },
            "type": "column",
          },
          ColumnElement {
            "name": "Column C",
            "options": Object {
              "id": "C",
            },
            "type": "column",
          },
        ],
        "options": Object {},
        "type": "header",
      },
      "options": Object {
        "name": "Example",
      },
      "rowsAndGroups": Array [
        RowElement {
          "cells": Array [
            CellElement {
              "options": Object {
                "id": "A",
              },
              "type": "cell",
              "value": "CellA",
            },
            CellElement {
              "options": Object {
                "id": "B",
              },
              "type": "cell",
              "value": "CellB",
            },
            CellElement {
              "options": Object {
                "id": "C",
              },
              "type": "cell",
              "value": "CellC",
            },
          ],
          "options": Object {},
          "type": "row",
        },
      ],
      "type": "worksheet",
    },
    WorksheetElement {
      "header": HeaderElement {
        "columns": Array [
          ColumnElement {
            "name": "Column A",
            "options": Object {
              "id": "A",
            },
            "type": "column",
          },
          ColumnElement {
            "name": "Column B",
            "options": Object {
              "id": "B",
            },
            "type": "column",
          },
          ColumnElement {
            "name": "Column C",
            "options": Object {
              "id": "C",
            },
            "type": "column",
          },
        ],
        "options": Object {},
        "type": "header",
      },
      "options": Object {
        "name": "Example2",
      },
      "rowsAndGroups": Array [
        RowElement {
          "cells": Array [
            CellElement {
              "options": Object {
                "id": "A",
              },
              "type": "cell",
              "value": "CellA",
            },
            CellElement {
              "options": Object {
                "id": "B",
              },
              "type": "cell",
              "value": "CellB",
            },
            CellElement {
              "options": Object {
                "id": "C",
              },
              "type": "cell",
              "value": "CellC",
            },
          ],
          "options": Object {},
          "type": "row",
        },
      ],
      "type": "worksheet",
    },
  ],
}
`;

exports[`worksheets should create worksheet with row and cells 1`] = `
WorksheetElement {
  "header": HeaderElement {
    "columns": Array [
      ColumnElement {
        "name": "Column A",
        "options": Object {
          "id": "A",
        },
        "type": "column",
      },
      ColumnElement {
        "name": "Column B",
        "options": Object {
          "id": "B",
        },
        "type": "column",
      },
      ColumnElement {
        "name": "Column C",
        "options": Object {
          "id": "C",
        },
        "type": "column",
      },
    ],
    "options": Object {},
    "type": "header",
  },
  "options": Object {
    "name": "Example",
  },
  "rowsAndGroups": Array [
    RowElement {
      "cells": Array [
        CellElement {
          "options": Object {
            "id": "A",
          },
          "type": "cell",
          "value": "CellA",
        },
        CellElement {
          "options": Object {
            "id": "B",
          },
          "type": "cell",
          "value": "CellB",
        },
        CellElement {
          "options": Object {
            "id": "C",
          },
          "type": "cell",
          "value": "CellC",
        },
      ],
      "options": Object {},
      "type": "row",
    },
  ],
  "type": "worksheet",
}
`;
